services/gitlab/gitlab-coverage.service.js

```diff
@@ -0,0 +1,143 @@
+    .regex(/^([0-9]+\.[0-9]+%)|unknown$/)
+    .required(),
+}).required()
+
+const queryParamSchema = Joi.object({
+  gitlab_url: optionalUrl,
+  job_name: Joi.string(),
+}).required()
+
+const documentation = `
+<p>
+  Important: If your project is publicly visible, but the badge is like this:
+  <img src="https://img.shields.io/badge/coverage-not&nbsp;set&nbsp;up-red" alt="coverage not set up"/>
+</p>
+<p>
+  Check if your pipelines are publicly visible as well.<br />
+  Navigate to your project settings on GitLab and choose General Pipelines under CI/CD.<br />
+  Then tick the setting Public pipelines.
+</p>
+<p>
+  Now your settings should look like this:
+</p>
+<img src="https://user-images.githubusercontent.com/12065866/67156911-e225a180-f324-11e9-93ad-10aafbb3e69e.png" alt="Setting Public pipelines set"/>
+<p>
+Also make sure you have set up code covrage parsing as described <a href="https://docs.gitlab.com/ee/ci/pipelines/settings.html#test-coverage-parsing">here</a>
+</p>
+<p>
+  Your badge should be working fine now.
+</p>
+`
+
+module.exports = class GitlabCoverage extends BaseSvgScrapingService {
+  static get category() {
+    return 'coverage'
+  }
+
+  static get route() {
+    return {
+      base: 'gitlab/coverage',
+      pattern: ':user/:repo/:branch',
+      queryParamSchema,
+    }
+  }
+
+  static get examples() {
+    return [
+      {
+        title: 'Gitlab code coverage',
+        namedParams: {
+          user: 'gitlab-org',
+          repo: 'gitlab-runner',
+          branch: 'master',
+        },
+        staticPreview: this.render({ coverage: 67 }),
+        documentation,
+      },
+      {
+        title: 'Gitlab code coverage (specific job)',
+        namedParams: {
+          user: 'gitlab-org',
+          repo: 'gitlab',
+          branch: 'master',
+        },
+        queryParams: { job_name: 'rspec:coverage' },
+        staticPreview: this.render({ coverage: 96 }),
+        documentation,
+      },
+      {
+        title: 'Gitlab code coverage (self-hosted)',
+        namedParams: { user: 'GNOME', repo: 'libhandy', branch: 'master' },
+        queryParams: { gitlab_url: 'https://gitlab.gnome.org' },
+        staticPreview: this.render({ coverage: 93 }),
+        documentation,
+      },
+      {
+        title: 'Gitlab code coverage (self-hosted, specific job)',
+        namedParams: { user: 'GNOME', repo: 'libhandy', branch: 'master' },
+        queryParams: {
+          gitlab_url: 'https://gitlab.gnome.org',
+          job_name: 'unit-test',
+        },
+        staticPreview: this.render({ coverage: 93 }),
+        documentation,
+      },
+    ]
+  }
+
+  static get defaultBadgeData() {
+    return { label: 'coverage' }
+  }
+
+  static render({ coverage }) {
+    return {
+      message: `${coverage.toFixed(0)}%`,
+      color: coveragePercentage(coverage),
+    }
+  }
+
+  async fetch(
+    { user, repo, branch },
+    { gitlab_url: baseUrl = 'https://gitlab.com', job_name }
+  ) {
```

Comment:
nit but we can use a single destructured object param in the signature here. the handle function has them split in two due to how the routing/server infra handles route vs. query params and passes them to these service classes, but we don't have to mimic that between these internal functions
```suggestion
  async fetch({
    user,
    repo,
    branch,
    gitlab_url: baseUrl = 'https://gitlab.com',
    job_name,
  }) {
```